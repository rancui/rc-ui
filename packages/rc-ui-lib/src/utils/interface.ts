import type { CSSProperties } from 'react';

export interface BaseTypeProps {
  style?: CSSProperties;
  className?: string;
  children?: React.ReactNode;
}

/** 指定挂载的节点 */
export type TeleportType = HTMLElement | (() => HTMLElement) | null;
/** 带有displayName的组件 */
export type WithDisplayNameReactElement = React.ReactElement & { type: { displayName: string } };
