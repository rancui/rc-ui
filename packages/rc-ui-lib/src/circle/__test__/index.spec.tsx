import * as React from 'react';
import { cleanup, render } from '@testing-library/react';
import { sleep } from '../../../tests/utils';
import { Circle } from '..';

describe('Circle', () => {
  afterEach(() => {
    cleanup();
  });

  it('should update to final rate immediately if speed is 0', async () => {
    const onChange = jest.fn();
    render(<Circle rate={50} speed={0} onChange={onChange} />);
    // await wrapper.setProps({ rate: 150 });
    // expect(wrapper.props().rate).toEqual(150);
    await sleep(50);
    expect(onChange).toHaveBeenCalledWith(50);
  });

  it('should render rate correctly when rate is changed', async () => {
    const { container, rerender } = render(<Circle rate={50} defaultRate={0} />);
    // expect(wrapper.props().rate).toEqual(50);
    expect(container).toMatchSnapshot();
    // await wrapper.setProps({ rate: 150 });
    // expect(wrapper.props().rate).toEqual(150);
    rerender(<Circle rate={150} defaultRate={0} />);
    expect(container).toMatchSnapshot();
  });

  it('should change circle size when using size prop', async () => {
    const { container, rerender } = render(<Circle defaultRate={0} size={100} />);
    const style = getComputedStyle(container.querySelector('.rc-circle'));
    expect(style.width).toEqual('100px');
    expect(style.height).toEqual('100px');

    rerender(<Circle defaultRate={0} size={150} />);
    const style2 = getComputedStyle(container.querySelector('.rc-circle'));
    expect(style2.width).toEqual('150px');
    expect(style2.height).toEqual('150px');
    expect(container).toMatchSnapshot();
  });

  it('should change stroke linecap when using stroke-linecap prop', async () => {
    const { container, rerender } = render(<Circle defaultRate={0} strokeLinecap="square" />);
    const style = getComputedStyle(container.querySelector('.rc-circle__hover'));
    expect(style.strokeLinecap).toEqual('square');

    rerender(<Circle defaultRate={0} strokeLinecap="butt" />);
    const style2 = getComputedStyle(container.querySelector('.rc-circle__hover'));
    expect(style2.strokeLinecap).toEqual('butt');
    expect(container).toMatchSnapshot();
  });

  it('should render start-position prop correctly', async () => {
    const { container, rerender } = render(<Circle startPosition="top" />);
    const style = getComputedStyle(container.querySelector('svg'));
    expect(style.transform).toEqual('');

    rerender(<Circle startPosition="right" />);
    const style2 = getComputedStyle(container.querySelector('svg'));
    expect(style2.transform).toEqual('rotate(90deg)');

    rerender(<Circle startPosition="bottom" />);
    const style3 = getComputedStyle(container.querySelector('svg'));
    expect(style3.transform).toEqual('rotate(180deg)');

    rerender(<Circle startPosition="left" />);
    const style4 = getComputedStyle(container.querySelector('svg'));
    expect(style4.transform).toEqual('rotate(270deg)');
  });

  it('should render text correctly', async () => {
    const { container } = render(<Circle defaultRate={70} text="70%" />);
    expect(container.querySelector('.rc-circle__text').textContent).toEqual('70%');
    expect(container).toMatchSnapshot();
  });

  it('should render linearGradient correctly', async () => {
    const { container } = render(
      <Circle
        defaultRate={70}
        color={{
          '0%': '#3fecff',
          '100%': '#6149f6',
        }}
      />,
    );
    // expect(container.querySelector('stop').props().stopColor).toEqual('#3fecff');
    // expect(container.querySelector('stop').at(1).props().stopColor).toEqual('#6149f6');
    expect(container).toMatchSnapshot();
  });

  it('should render clockwise and anticlockwise correctly', async () => {
    const { container, rerender } = render(<Circle defaultRate={70} clockwise text="70%" />);
    // expect(container.querySelector('svg').find('path').at(1).props().d.substr(-10, 1)).toEqual('1');
    expect(container).toMatchSnapshot();

    rerender(<Circle defaultRate={70} clockwise={false} text="70%" />);
    // expect(container.querySelector('svg').find('path').at(1).props().d.substr(-10, 1)).toEqual('0');
    expect(container).toMatchSnapshot();
  });

  it('should render layerColor correctly', () => {
    const { container } = render(<Circle defaultRate={70} layerColor="#ebedf0" />);
    const style = getComputedStyle(container.querySelector('.rc-circle__layer'));
    expect(style.stroke).toEqual('#ebedf0');
    expect(container).toMatchSnapshot();
  });

  it('should render fill prop correctly', () => {
    const { container } = render(<Circle defaultRate={70} fill="#ff0000" />);
    const style = getComputedStyle(container.querySelector('.rc-circle__layer'));
    expect(style.fill).toEqual('#ff0000');
    expect(container).toMatchSnapshot();
  });

  it('hould emit onChange event during animation', async () => {
    const onChange = jest.fn();
    render(<Circle speed={150} rate={150} onChange={onChange} />);
    await sleep(50);
    expect(onChange).toHaveBeenCalled();
  });
});
