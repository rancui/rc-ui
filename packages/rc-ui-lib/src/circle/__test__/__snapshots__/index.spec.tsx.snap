// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Circle should change circle size when using size prop 1`] = `
<div>
  <div
    class="rc-circle"
    style="width: 150px; height: 150px;"
  >
    <svg
      viewBox="0 0 1040 1040"
    >
      <path
        class="rc-circle__layer"
        d="M 520 520 m 0, -500 a 500, 500 0 1, 1 0, 1000 a 500, 500 0 1, 1 0, -1000"
        style="fill: none; stroke-width: 40px;"
      />
      <path
        class="rc-circle__hover"
        d="M 520 520 m 0, -500 a 500, 500 0 1, 1 0, 1000 a 500, 500 0 1, 1 0, -1000"
        style="stroke-width: 41px; stroke-dasharray: 0px 3140px;"
      />
    </svg>
  </div>
</div>
`;

exports[`Circle should change stroke linecap when using stroke-linecap prop 1`] = `
<div>
  <div
    class="rc-circle"
  >
    <svg
      viewBox="0 0 1040 1040"
    >
      <path
        class="rc-circle__layer"
        d="M 520 520 m 0, -500 a 500, 500 0 1, 1 0, 1000 a 500, 500 0 1, 1 0, -1000"
        style="fill: none; stroke-width: 40px;"
      />
      <path
        class="rc-circle__hover"
        d="M 520 520 m 0, -500 a 500, 500 0 1, 1 0, 1000 a 500, 500 0 1, 1 0, -1000"
        style="stroke-width: 41px; stroke-linecap: butt; stroke-dasharray: 0px 3140px;"
      />
    </svg>
  </div>
</div>
`;

exports[`Circle should render clockwise and anticlockwise correctly 1`] = `
<div>
  <div
    class="rc-circle"
  >
    <svg
      viewBox="0 0 1040 1040"
    >
      <path
        class="rc-circle__layer"
        d="M 520 520 m 0, -500 a 500, 500 0 1, 1 0, 1000 a 500, 500 0 1, 1 0, -1000"
        style="fill: none; stroke-width: 40px;"
      />
      <path
        class="rc-circle__hover"
        d="M 520 520 m 0, -500 a 500, 500 0 1, 1 0, 1000 a 500, 500 0 1, 1 0, -1000"
        style="stroke-width: 41px; stroke-dasharray: 2198px 3140px;"
      />
    </svg>
    <div
      class="rc-circle__text"
    >
      70%
    </div>
  </div>
</div>
`;

exports[`Circle should render clockwise and anticlockwise correctly 2`] = `
<div>
  <div
    class="rc-circle"
  >
    <svg
      viewBox="0 0 1040 1040"
    >
      <path
        class="rc-circle__layer"
        d="M 520 520 m 0, -500 a 500, 500 0 1, 0 0, 1000 a 500, 500 0 1, 0 0, -1000"
        style="fill: none; stroke-width: 40px;"
      />
      <path
        class="rc-circle__hover"
        d="M 520 520 m 0, -500 a 500, 500 0 1, 0 0, 1000 a 500, 500 0 1, 0 0, -1000"
        style="stroke-width: 41px; stroke-dasharray: 2198px 3140px;"
      />
    </svg>
    <div
      class="rc-circle__text"
    >
      70%
    </div>
  </div>
</div>
`;

exports[`Circle should render fill prop correctly 1`] = `
<div>
  <div
    class="rc-circle"
  >
    <svg
      viewBox="0 0 1040 1040"
    >
      <path
        class="rc-circle__layer"
        d="M 520 520 m 0, -500 a 500, 500 0 1, 1 0, 1000 a 500, 500 0 1, 1 0, -1000"
        style="fill: #ff0000; stroke-width: 40px;"
      />
      <path
        class="rc-circle__hover"
        d="M 520 520 m 0, -500 a 500, 500 0 1, 1 0, 1000 a 500, 500 0 1, 1 0, -1000"
        style="stroke-width: 41px; stroke-dasharray: 2198px 3140px;"
      />
    </svg>
  </div>
</div>
`;

exports[`Circle should render layerColor correctly 1`] = `
<div>
  <div
    class="rc-circle"
  >
    <svg
      viewBox="0 0 1040 1040"
    >
      <path
        class="rc-circle__layer"
        d="M 520 520 m 0, -500 a 500, 500 0 1, 1 0, 1000 a 500, 500 0 1, 1 0, -1000"
        style="fill: none; stroke: #ebedf0; stroke-width: 40px;"
      />
      <path
        class="rc-circle__hover"
        d="M 520 520 m 0, -500 a 500, 500 0 1, 1 0, 1000 a 500, 500 0 1, 1 0, -1000"
        style="stroke-width: 41px; stroke-dasharray: 2198px 3140px;"
      />
    </svg>
  </div>
</div>
`;

exports[`Circle should render linearGradient correctly 1`] = `
<div>
  <div
    class="rc-circle"
  >
    <svg
      viewBox="0 0 1040 1040"
    >
      <defs>
        <lineargradient
          class="linearGradient"
          id="rc-circle-13"
          x1="100%"
          x2="0%"
          y1="0%"
          y2="0%"
        >
          <stop
            offset="0%"
            stop-color="#3fecff"
          />
          <stop
            offset="100%"
            stop-color="#6149f6"
          />
        </lineargradient>
      </defs>
      <path
        class="rc-circle__layer"
        d="M 520 520 m 0, -500 a 500, 500 0 1, 1 0, 1000 a 500, 500 0 1, 1 0, -1000"
        style="fill: none; stroke-width: 40px;"
      />
      <path
        class="rc-circle__hover"
        d="M 520 520 m 0, -500 a 500, 500 0 1, 1 0, 1000 a 500, 500 0 1, 1 0, -1000"
        stroke="url(#rc-circle-13)"
        style="stroke: url(#rc-circle-13); stroke-width: 41px; stroke-dasharray: 2198px 3140px;"
      />
    </svg>
  </div>
</div>
`;

exports[`Circle should render rate correctly when rate is changed 1`] = `
<div>
  <div
    class="rc-circle"
  >
    <svg
      viewBox="0 0 1040 1040"
    >
      <path
        class="rc-circle__layer"
        d="M 520 520 m 0, -500 a 500, 500 0 1, 1 0, 1000 a 500, 500 0 1, 1 0, -1000"
        style="fill: none; stroke-width: 40px;"
      />
      <path
        class="rc-circle__hover"
        d="M 520 520 m 0, -500 a 500, 500 0 1, 1 0, 1000 a 500, 500 0 1, 1 0, -1000"
        style="stroke-width: 41px; stroke-dasharray: 0px 3140px;"
      />
    </svg>
  </div>
</div>
`;

exports[`Circle should render rate correctly when rate is changed 2`] = `
<div>
  <div
    class="rc-circle"
  >
    <svg
      viewBox="0 0 1040 1040"
    >
      <path
        class="rc-circle__layer"
        d="M 520 520 m 0, -500 a 500, 500 0 1, 1 0, 1000 a 500, 500 0 1, 1 0, -1000"
        style="fill: none; stroke-width: 40px;"
      />
      <path
        class="rc-circle__hover"
        d="M 520 520 m 0, -500 a 500, 500 0 1, 1 0, 1000 a 500, 500 0 1, 1 0, -1000"
        style="stroke-width: 41px; stroke-dasharray: 0px 3140px;"
      />
    </svg>
  </div>
</div>
`;

exports[`Circle should render text correctly 1`] = `
<div>
  <div
    class="rc-circle"
  >
    <svg
      viewBox="0 0 1040 1040"
    >
      <path
        class="rc-circle__layer"
        d="M 520 520 m 0, -500 a 500, 500 0 1, 1 0, 1000 a 500, 500 0 1, 1 0, -1000"
        style="fill: none; stroke-width: 40px;"
      />
      <path
        class="rc-circle__hover"
        d="M 520 520 m 0, -500 a 500, 500 0 1, 1 0, 1000 a 500, 500 0 1, 1 0, -1000"
        style="stroke-width: 41px; stroke-dasharray: 2198px 3140px;"
      />
    </svg>
    <div
      class="rc-circle__text"
    >
      70%
    </div>
  </div>
</div>
`;
